[CI] AsyncSearchActionIT testRestartAfterCompletion failing

**Build Scans:**
- [elasticsearch-intake #42080 / part2](https://gradle-enterprise.elastic.co/s/glcecbj3ifkh6)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:async-search:internalClusterTest" --tests "org.elasticsearch.xpack.search.AsyncSearchActionIT.testRestartAfterCompletion" -Dtests.seed=EAA8AFA7CAF822FE -Dtests.locale=no-NO -Dtests.timezone=Asia/Barnaul -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.search.AsyncSearchActionIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testRestartAfterCompletion'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: There are still [1] in-flight contexts. The first one's creation site is listed as the cause of this exception.
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testRestartAfterCompletion
- [main] 1 consecutive failures in step part2
- [main] 1 consecutive failures in pipeline elasticsearch-intake

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -644,12 +644,6 @@ tests:
 - class: org.elasticsearch.upgrades.MlMappingsUpgradeIT
   method: testMappingsUpgrade
   issue: https://github.com/elastic/elasticsearch/issues/152115
-- class: org.elasticsearch.xpack.search.AsyncSearchActionIT
-  method: testRestartAfterCompletion
-  issue: https://github.com/elastic/elasticsearch/issues/152116
-- class: org.elasticsearch.xpack.search.AsyncSearchActionIT
-  method: testSearchPhaseFailureLeak
-  issue: https://github.com/elastic/elasticsearch/issues/152117
 - class: org.elasticsearch.index.engine.InternalEngineTests
   method: testLookupSeqNoByIdInLucene
   issue: https://github.com/elastic/elasticsearch/issues/152120

--- a/x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/AsyncSearchIntegTestCase.java
+++ b/x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/AsyncSearchIntegTestCase.java
@@ -154,12 +154,13 @@ public abstract class AsyncSearchIntegTestCase extends ESIntegTestCase {
         ensureAllSearchContextsReleased();
 
         internalCluster().restartNode(node.getName(), new InternalTestCluster.RestartCallback() {});
-        // Ensure shards are ready before unpausing the maintenance service. If we unpaused first,
-        // the maintenance service could dispatch a DeleteByQueryRequest to a shard that is still
-        // INITIALIZING, hitting the waitForSearchReady path. The resulting reader context would be
-        // registered after ensureAllSearchContextsReleased() passes, causing a spurious
-        // assertNoInFlightContext() failure in subsequent-test teardown.
+        // The restarted node gets a new, unpaused maintenance service that the pre-restart pause did
+        // not cover. It can dispatch a DeleteByQueryRequest to a still-INITIALIZING shard, registering
+        // an async waitForSearchReady callback that opens a reader context only after the shard starts.
+        // Wait for shards to start, then pause and drain any such in-flight contexts before unpausing.
         ensureYellow(ASYNC_RESULTS_INDEX, indexName);
+        pauseMaintenanceService();
+        ensureAllSearchContextsReleased();
         unpauseMaintenanceService();
     }
 

